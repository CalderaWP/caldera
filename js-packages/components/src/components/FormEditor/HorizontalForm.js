import React, {Fragment} from 'react';
import PropTypes from 'prop-types';
import {CalderaForm} from "@calderawp/forms";

export const HorizontalForm = (
	{
		fields,
		initialValues,
		onClose,
		onChange,
		onBlur,
		instanceId
	}) => {
	const rows = [];
	let i = 0;
	fields.forEach(field => {
		rows.push({
			rowId: `${instanceId}-${i}`,
			columns: [
				{
					columnId: `c-${instanceId}-${i}`,
					fields: [field],
					width: '1'
				}
			]
		});
		i++;
	});
	return (
		<Fragment>
			<CalderaForm
				formRows={rows}
				initialValues={initialValues}
				onSubmit={onClose}
				onChange={onChange}
				onBlur={onBlur}
			/>

		</Fragment>
	);
};



HorizontalForm.propTypes = {
	fields: PropTypes.array,
	initialValues: PropTypes.object,
	onSubmit: PropTypes.func,
	onChange: PropTypes.func,
	onBlur: PropTypes.func,
	onClose: PropTypes.func,
	initialValues: PropTypes.object,
};

const _noop = () => {};
HorizontalForm.defaultProps = {
	fields: {},
	onSubmit: _noop,
	onChange: _noop,
	onBlur: _noop,
	initialValues: {}
};


