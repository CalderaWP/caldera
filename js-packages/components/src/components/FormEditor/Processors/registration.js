

const defaultRegistry = [];

/**
 * Add a processor type to form editor
 *
 * @param {*} processorType Defintion of processor type
 * @param {Array} registry Optional. Collection to add to -- generally you leave this at default.
 * @return {Array}
 */
export const registerProcessorType = (processorType,registry = defaultRegistry ) => {
	if( 'object' === typeof  processorType && processorType.hasOwnProperty('type' ) ) {
		return [ processorType,...registry];
	}

	return registry;

};

export const unregisterProcessorType = (processorType,registry = defaultRegistry ) => {
	if( 'object' === typeof  processorType && processorType.hasOwnProperty('type' ) ) {
		processorType = processorType.type;
	}
	const index = registry.findIndex( p => p.type === processorType );
	if( 0<= index ){
		registry.splice(index,1);
		return [ ...registry];
	}
	return registry;
};
