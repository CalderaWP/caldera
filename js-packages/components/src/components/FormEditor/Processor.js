import React, {Fragment, Component} from 'react';
import {HorizontalForm} from './HorizontalForm';
import {TabPanel} from '@wordpress/components';
import { Row,Column } from '@calderawp/factory';

import PropTypes from 'prop-types';

export class Processor extends Component {


	state = {
		activeTab: 'settings',
	};

	onSetTab = (activeTab) => {
		this.setState({activeTab});
	};

	render() {
		const {onClose, onRemove} = this.props;
		return (
			<Fragment>
				<Row>
					<TabPanel
						className="caldera-processor"
						activeClass="active-tab"
						onSelect={this.onSetTab}
						initialTabName={'settings'}
						tabs={[
							{
								name: 'settings',
								title: 'Settings',
								className: 'caldera-processor-settings-tab-btn',
							},
							{
								name: 'conditionals',
								title: 'Conditionals',
								className: 'caldera-processor-conditionals-tab-btn',
							},
						]}>
						{
							(tab) => {
								const {name} = tab;
								if( 'settings' === name ){
									return <HorizontalForm {...this.props} className={'caldera-processor-settings'}/>
								}
								return <div className={'caldera-processor-conditionals'}>Conditionals</div>
							}
						}
					</TabPanel>
				</Row>
				<Row>
					<Column
						width={'1/2'}
					>
						<button
							onClick={onClose}
						>
							Close
						</button>
					</Column>
					<Column
						width={'1/2'}
					>
						<button
							onClick={onRemove}
						>
							Remove
						</button>
					</Column>
				</Row>
			</Fragment>
		)
	}
}

Processor.propTypes = {
	...HorizontalForm.propTypes,
	initialActiveTab: PropTypes.string,
	onRemove: PropTypes.func,
};
Processor.defaultProps = {
	...HorizontalForm.defaultProps,
	initialActiveTab: 'settings'
};
