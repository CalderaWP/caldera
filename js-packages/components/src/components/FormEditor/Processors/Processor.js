import React, { Fragment, Component } from 'react';
import { HorizontalForm } from '../../HorizontalForm/HorizontalForm';
import { TabPanel } from '@wordpress/components';
import { Row, Column } from '@calderawp/factory';

import PropTypes from 'prop-types';

export class Processor extends Component {
	state = {
		activeTab: 'settings'
	};

	onSetTab = activeTab => {
		this.setState({ activeTab });
	};

	tabs = [
		{
			name: 'settings',
			title: 'Settings',
			className: 'caldera-processor-settings-tab-btn'
		},
		{
			name: 'conditionals',
			title: 'Conditionals',
			className:
				'caldera-processor-conditionals-tab-btn'
		}
	];
	render() {
		const { onClose, onRemove } = this.props;
		return (
			<Fragment>
				<Row>
					<TabPanel
						className="caldera-processor"
						activeClass="active-tab"
						onSelect={this.onSetTab}
						initialTabName={'settings'}
						tabs={this.tabs}
					>
						{() => {
							const {activeTab} = this.state;
							const tab =  this.tabs.find( t => t.name === activeTab );
							const { name } = tab;
							if ('settings' === name) {
								return (
									<HorizontalForm
										{...this.props}
										className={'caldera-processor-settings'}
									/>
								);
							}
							return (
								<div
									className={'caldera-processor-conditionals'}
								>
									Conditionals
								</div>
							);
						}}
					</TabPanel>
				</Row>
				<Row>
					<Column width={'1/2'}>
						<button
							className={'caldera-processor-close'}
							onClick={onClose}>
							Close
						</button>
					</Column>
					<Column width={'1/2'}>
						<button
							className={'caldera-processor-remove'}
							onClick={onRemove}>
							Remove
						</button>
					</Column>
				</Row>
			</Fragment>
		);
	}
}

//should have label!
Processor.propTypes = {
	...HorizontalForm.propTypes,
	initialActiveTab: PropTypes.string,
	onRemove: PropTypes.func,
	type: PropTypes.string,
	id: PropTypes.string,
};

Processor.defaultProps = {
	...HorizontalForm.defaultProps,
	initialActiveTab: 'settings'
};
