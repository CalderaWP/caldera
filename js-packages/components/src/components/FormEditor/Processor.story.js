import React from 'react';
import { storiesOf } from '@storybook/react';
import { Processor } from './Processor';

import {
	checkboxFieldset,
	selectField,
	checkboxField,
	numberField,
	textField,
	emailField,
	radioField
} from '@calderawp/factory';
storiesOf('Processor', module).add('Works with a bunch of fields', () => (
	<Processor
		fields={[
			checkboxFieldset,
			selectField,
			checkboxField,
			numberField,
			textField,
			emailField,
			radioField
		]}
		onChange={values => console.log(values)}
		onClose={values => console.log(values)}
		onRemove={values => console.log(values)}
	/>
));
