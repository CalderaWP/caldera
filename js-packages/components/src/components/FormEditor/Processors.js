import React, {Component, Fragment} from 'react';
import PropTypes from 'prop-types';
import {Processor} from "./Processor";

export class Processors extends Component {

	state = {
		activeProcessorId: ''
	};

	/**
	 * Set which processor is active
	 * @param activeProcessorId
	 */
	setActive = (activeProcessorId) => {
		this.setState({activeProcessorId});
	};

	/**
	 * Check if a processor is active by ID
	 * @param processorId
	 * @return {string|boolean}
	 */
	isActiveProcessor = (processorId) => {
		const {activeProcessorId} = this.state;
		return activeProcessorId && processorId === activeProcessorId;
	};

	/**
	 * When a change in value for a processor is received, send all processors up.
	 *
	 * @param processorId
	 * @param updatedProcessor
	 */
	handleProcessorChange = (processorId, updatedProcessor) => {
		const {processors, updateProcessors} = this.props;
		let processorIndex = processors.findIndex(processor => {
			return processorId === processor.id
		});
		if (-1 !== processorIndex) {
			let processor = processors[processorIndex];
			let update = [
				...processors.splice(processorIndex - 1, 1),
				{
					...processor,
					...updatedProcessor
				}
			];
			updateProcessors(update);
		}


	};

	/**
	 * When a processor is removed, send updated list to change handler.
	 *
	 * @param processorId
	 */
	handleRemoveProcessor = (processorId) => {
		const {processors, updateProcessors} = this.props;
		updateProcessors([...processors.filter(processor => {
			return processorId !== processor.id
		})]);
	};

	/** @inheritDoc **/
	render() {
		const {activeProcessor} = this.state;
		const {processors} = this.props;
		return (
			<div>
				<div>
					{processors.map(processor => {
						const {
							id,
							fields,
							config,
							type,
							label
						} = processor;
						if (this.isActiveProcessor(id)) {

							return (
								<Fragment key={id}>
									<Processor
										className={`caldera-forms-active-processor-${id}`}
										fields={fields}
										initialValues={config}
										label={label}
										type={type}
										onChange={(fieldValues) => this.handleProcessorChange(id, fieldValues)}
										onRemove={() => this.handleRemoveProcessor(id)}
										onClose={() => this.setActive('')}
									/>
								</Fragment>
							)
						}
						return (
							<Fragment>
								<button
									className={`caldera-forms-choose-processor caldera-forms-choose-processor-${id}`}
									onClick={() => this.setActive(id)}
								>
									{label ? label : type}
								</button>
							</Fragment>

						)
					})}
				</div>
			</div>
		);
	}
}

export const processorsCollectionPropType = PropTypes.arrayOf(
	PropTypes.shape({
		id: PropTypes.string,
		label: PropTypes.string,
		type: PropTypes.string,
		fields: PropTypes.array,
		config: PropTypes.object
	})
);

Processors.propTypes = {
	processors: processorsCollectionPropType,
	updateProcessors: PropTypes.func,
	form: PropTypes.object,
	formFields: PropTypes.object
};

Processors.defaultProps = {
	processors: []
};
