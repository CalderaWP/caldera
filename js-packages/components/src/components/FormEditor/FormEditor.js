import React, {Component, Fragment,createElement} from 'react';
import PropTypes from 'prop-types';
import {Processors,processorsCollectionPropType} from './Processors';
import { TabPanel } from '@wordpress/components';
import { Row, Column } from '@calderawp/factory';
import {processorsCollection} from "./processors.fixtures";
import {fieldAreaFactory} from '@calderawp/factory';

export class FormEditor extends Component {
	state = {
		activeTab: 'processors',
	};

	onSetTab = activeTab => {
		this.setState({ activeTab });
	};

	updateProcessors = processors => {
		const {updateForm,form} = this.props;
		updateForm({
			...form,
			processors
		});
	};

	getFormProcessors = () => {
		const {form} = this.props;
		return form.processors ? form.processors : [];
	};

	getFormFields = () => {
		const {form} = this.props;
		return form.fields ? form.fields : [];
	};

	render() {
		const {form,processorTypes,updateForm} = this.props;
		return (
			<div>
				<Row>
					<Column>
						<p>{form.name}</p>
					</Column>
				</Row>
				<TabPanel
					className="caldera-processor"
					activeClass="active-tab"
					onSelect={this.onSetTab}
					initialTabName={'processors'}
					tabs={[
						{
							name: 'editor',
							title: 'Layout',
							className: 'caldera-forms-editor-layout-tab-btn',
							classNameForComponent: 'caldera-forms-editor-layout'
						},
						{
							name: 'processors',
							title: 'Processors',
							className: 'caldera-forms-editor-processors-tab-btn',
							classNameForComponent: 'caldera-forms-editor-processors'

						},
						{
							name: 'Mailer',
							title: 'mailer',
							className: 'caldera-forms-editor-mailer-tab-btn',
							classNameForComponent: 'caldera-forms-editor-mailer'
						},
						{
							name: 'settings',
							title: 'Settings',
							className: 'caldera-forms-editor-settings-tab-btn',
							classNameForComponent: 'caldera-forms-editor-settings'

						},
					]}
				>
					{tab => {
						const { name,classNameForComponent} = tab;
						if ('processors' === name) {
							return (
								<Processors
									processors={this.getFormProcessors()}
									form={form}
									formFields={this.getFormFields()}
									updateProcessors={this.updateProcessors}
								/>

							);
						}
						if( 'settings' === name ) {
							const nameField = {
								fieldType: 'text',
								value: form.name,
								label: 'Form Name',
								fieldId: 'formName',
								required: true,
							};
							return (
								<Fragment>
									{fieldAreaFactory(
										nameField,
										(name) => {
											updateForm({
												...form,
												name
											})
										}
									)}
								</Fragment>
							)
						}
						return (
							<div
								className={classNameForComponent}
							>
								{tab.title}
							</div>
						);
					}}
				</TabPanel>
			</div>
		);
	}
}

FormEditor.propTypes = {
	processorsTypes: PropTypes.arrayOf(PropTypes.shape({
		type: PropTypes.string,
	})),
	updateForm: PropTypes.func,
	form: PropTypes.shape({
		id: PropTypes.string,
		name: PropTypes.string,
		fields: PropTypes.array,
		conditionals: PropTypes.array,
		settings: PropTypes.array,
		processors: processorsCollectionPropType,
	}),
};

