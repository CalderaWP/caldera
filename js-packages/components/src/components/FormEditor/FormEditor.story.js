import React, {Component} from 'react';
import { storiesOf } from '@storybook/react';
import { Processor } from './Processor';
import { Processors } from './Processors';
import { HorizontalForm } from './HorizontalForm';
import { FormEditor } from './FormEditor';
import {processorsCollection} from './processors.fixtures';
import {
	checkboxFieldset,
	selectField,
	checkboxField,
	numberField,
	textField,
	emailField,
	radioField
} from '@calderawp/factory';
storiesOf('FormEditor', module).add('The processor', () => (
	<Processor
		fields={[
			checkboxFieldset,
			selectField,
			checkboxField,
			numberField,
			textField,
			emailField,
			radioField
		]}
		type={'mailchimp'}
		label={'Main Segment'}
		onChange={values => console.log(values)}
		onClose={values => console.log(values)}
		onRemove={values => console.log(values)}
	/>
));

storiesOf('FormEditor', module).add('The processors list', () => (
	<Processors
		processors={processorsCollection}
		onChange={values => console.log(values)}
		onClose={values => console.log(values)}
		onRemove={values => console.log(values)}
		formFields={[]}
		form={ {id: 'Name'}}
	/>
));

const form = {
	id: 'a form',
	processors: processorsCollection,
};

const processorTypes = [
	{
		type: 'apiRequest'
	},
	{
		type: 'redirect'
	}
];

class MockFormApp extends Component {
	state = {
		form: {
			id: 'a-form',
			name: 'Form Name',
			processors: processorsCollection,
		}
	};

	updateForm = (form) => this.setState({form});

	render(){
		const {form} = this.state;
		return(
			<FormEditor
				processorTypes={processorTypes}
				updateForm={this.updateForm}
				form={ form }
			/>
		)
	}
}

storiesOf('FormEditor', module).add('The form editor', () => (
	<MockFormApp/>
));

storiesOf('HorizontalForm', module).add('Works with a bunch of fields', () => (
	<HorizontalForm
		fields={[
			checkboxFieldset,
			selectField,
			checkboxField,
			numberField,
			textField,
			emailField,
			radioField
		]}
		onChange={values => console.log(values)}
		onBlur={values => console.log(values)}
		onClose={values => console.log(values)}
	/>
));

