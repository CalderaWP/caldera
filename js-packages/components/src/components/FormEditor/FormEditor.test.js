import React, {Component} from 'react';
import PropTypes from 'prop-types';
import {FormEditor} from "./FormEditor";
import {Processors} from "./Processors";
import {mount} from 'enzyme';
import renderer from 'react-test-renderer';
import {processorsCollection} from './processors.fixtures';
const form = {
	id: 'a-form',
	name: 'Form Name',
	processors: processorsCollection,
};

const processorTypes = [
	{
		type: 'apiRequest'
	},
	{
		type: 'redirect'
	}
];

let updateForm = jest.fn();


describe('FormEditor', () => {

	beforeEach(() => {
		updateForm = jest.fn()
	});


	it('should match snapshot', () => {
		const component =
			renderer.create(<FormEditor
				processorTypes={processorTypes}
				updateForm={updateForm}
				form={ form }
			/>)
		expect(component.toJSON()).toMatchSnapshot();
	});
	it('Can switch tabs', () => {
		const component =
			mount(<FormEditor
				processorTypes={processorTypes}
				updateForm={updateForm}
				form={ form }
			/>);
		expect(component.find('.caldera-forms-editor-settings').length).toEqual(0);
		component.setState({activeTab: 'settings'});
		expect(component.find('.caldera-forms-editor-settings').length).toEqual(1);

	});


});

describe('Form editor settings', () => {
	beforeEach(() => {
		updateForm = jest.fn()
	});
	it('should update form name', () => {
		const component =
			mount(<FormEditor
				processorTypes={processorTypes}
				updateForm={updateForm}
				form={ form }
			/>);
		component.find( '.caldera-forms-editor-settings-tab-btn').simulate( 'click' );

		const expectedValue = 'Creatures';
		component.setState({activeTab: 'settings'});

		component.find( '#formName').simulate( 'change', event );
		expect(component.state('form').name).toEqual(expectedValue);

	});
});
