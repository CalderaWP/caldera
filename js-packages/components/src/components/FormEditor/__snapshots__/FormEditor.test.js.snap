// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormEditor should match snapshot 1`] = `
<div>
  <div
    className="sc-bdVaJa caldera-row sc-bwzfXH pjOvZ"
    id={undefined}
  >
    <div
      className="caldera-column sc-bdVaJa iHZvIS"
      id={undefined}
      width={undefined}
    >
      <p>
        Form Name
      </p>
    </div>
  </div>
  <div
    className="caldera-processor"
  >
    <div
      aria-orientation="horizontal"
      className="components-tab-panel__tabs"
      onFocus={undefined}
      onKeyDown={[Function]}
      role="tablist"
    >
      <button
        aria-controls="0-editor-view"
        aria-selected={false}
        className="caldera-forms-editor-layout-tab-btn "
        id="0-editor"
        onClick={[Function]}
        role="tab"
        tabIndex={-1}
      >
        Layout
      </button>
      <button
        aria-controls="0-processors-view"
        aria-selected={true}
        className="caldera-forms-editor-processors-tab-btn active-tab"
        id="0-processors"
        onClick={[Function]}
        role="tab"
        tabIndex={null}
      >
        Processors
      </button>
      <button
        aria-controls="0-Mailer-view"
        aria-selected={false}
        className="caldera-forms-editor-mailer-tab-btn "
        id="0-Mailer"
        onClick={[Function]}
        role="tab"
        tabIndex={-1}
      >
        mailer
      </button>
      <button
        aria-controls="0-settings-view"
        aria-selected={false}
        className="caldera-forms-editor-settings-tab-btn "
        id="0-settings"
        onClick={[Function]}
        role="tab"
        tabIndex={-1}
      >
        Settings
      </button>
    </div>
    <div
      aria-labelledby="0-processors"
      className="components-tab-panel__tab-content"
      id="0-processors-view"
      role="tabpanel"
      tabIndex="0"
    >
      <div
        className="caldera-forms-editor-processors"
      >
        <h2>
          Processors
        </h2>
        <div>
          <div>
            <div
              className="sc-bdVaJa caldera-row sc-bwzfXH pjOvZ"
              id={undefined}
            >
              <button
                className="caldera-forms-choose-processor caldera-forms-choose-processor-season7Episode2"
                onClick={[Function]}
              >
                Test sending form data to test API
              </button>
            </div>
            <div
              className="sc-bdVaJa caldera-row sc-bwzfXH pjOvZ"
              id={undefined}
            >
              <button
                className="caldera-forms-choose-processor caldera-forms-choose-processor-p2"
                onClick={[Function]}
              >
                Main MailChimp
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
